src/control/lib/hardware/fabric.go

```diff
@@ -0,0 +1,632 @@
+func (n *NUMAAffinityBuilder) BuildPart(ctx context.Context, fis *FabricInterfaceSet) error {
+	if n == nil {
+		return errors.New("NUMAAffinityBuilder is nil")
+	}
+
+	if fis == nil {
+		return errors.New("FabricInterfaceSet is nil")
+	}
+
+	devsByName := n.topo.AllDevices()
+
+	for _, name := range fis.Names() {
+		fi, err := fis.GetInterface(name)
+		if err != nil {
+			n.log.Errorf("can't update interface %s: %s", name, err.Error())
+			continue
+		}
+
+		if fi.DeviceClass == Loopback {
+			// Loopback is not a physical device
+			continue
+		}
+
+		dev, exists := devsByName[name]
+		if !exists {
+			n.log.Errorf("fabric interface %q not found in topology", name)
+			continue
+		}
+
+		fi.NUMANode = dev.NUMANode.ID
+	}
+	return nil
+}
+
+func newNUMAAffinityBuilder(log logging.Logger, topo *Topology) *NUMAAffinityBuilder {
+	return &NUMAAffinityBuilder{
+		log:  log,
+		topo: topo,
+	}
+}
+
+// NetDevClassBuilder is a builder updates FabricInterfaces with a NetDevClass.
+type NetDevClassBuilder struct {
+	log      logging.Logger
+	provider NetDevClassProvider
+}
+
+// BuildPart updates existing FabricInterface structures in the FabricInterfaceSet to include a
+// net device class (i.e. device type), if available.
+func (n *NetDevClassBuilder) BuildPart(ctx context.Context, fis *FabricInterfaceSet) error {
+	if n == nil {
+		return errors.New("NetDevClassBuilder is nil")
+	}
+
+	if fis == nil {
+		return errors.New("FabricInterfaceSet is nil")
+	}
+
+	for _, name := range fis.Names() {
+		fi, err := fis.GetInterface(name)
+		if err != nil {
+			return err
+		}
+
+		ndc, err := n.provider.GetNetDevClass(name)
+		if err != nil {
+			n.log.Error(err.Error())
+		}
+
+		fi.DeviceClass = ndc
+	}
+	return nil
+}
+
+func newNetDevClassBuilder(log logging.Logger, provider NetDevClassProvider) *NetDevClassBuilder {
+	return &NetDevClassBuilder{
+		log:      log,
+		provider: provider,
+	}
+}
+
+// FabricInterfaceSetBuilderConfig contains the configuration used by FabricInterfaceSetBuilders.
+type FabricInterfaceSetBuilderConfig struct {
+	Topology                 *Topology
+	FabricInterfaceProviders []FabricInterfaceProvider
+	NetDevClassProvider      NetDevClassProvider
+}
+
+func defaultFabricInterfaceSetBuilders(log logging.Logger, config *FabricInterfaceSetBuilderConfig) []FabricInterfaceSetBuilder {
+	return []FabricInterfaceSetBuilder{
+		newFabricInterfaceBuilder(log, config.FabricInterfaceProviders...),
+		newNetDevClassBuilder(log, config.NetDevClassProvider),
+		newOSDeviceBuilder(log, config.Topology),
+		newNUMAAffinityBuilder(log, config.Topology),
+	}
+}
+
+// FabricScannerConfig contains the parameters required to set up a FabricScanner.
+type FabricScannerConfig struct {
+	TopologyProvider         TopologyProvider
+	FabricInterfaceProviders []FabricInterfaceProvider
+	NetDevClassProvider      NetDevClassProvider
+}
+
+// IsValid checks if the FabricScannerConfig is valid.
+func (c *FabricScannerConfig) IsValid() bool {
```

Comment:
minor: Might be better to rename this to `Validate()` and have it return an error, so that it can specify exactly why the config is invalid.

Reply:
Good idea. Fixing.